src/snowflake/connector/pandas_tools.py

```diff
@@ -202,90 +192,85 @@ def write_pandas(
-        # columns in order
-        quote = '"' if quote_identifiers else ""
-        create_table_columns = ", ".join(
-            [f"{quote}{c}{quote} {column_type_mapping[c]}" for c in df.columns]
+        try:
+            column_type_mapping = dict(
+                cursor.execute(infer_schema_sql, _is_internal=True).fetchall()
+            )
+            # Infer schema can return the columns out of order depending on the chunking we do when uploading
+            # so we have to iterate through the dataframe columns to make sure we create the table with its
+            # columns in order
+            create_table_columns = ", ".join(
+                [f"{quote}{c}{quote} {column_type_mapping[c]}" for c in df.columns]
+            )
+        finally:
+            drop_object(file_format_name, "file format")
+
+        target_table_name = (
+            f"{location}{quote}{random_name() if overwrite else table_name}{quote}"
         )
         create_table_sql = (
-            f"CREATE {table_type.upper()} TABLE IF NOT EXISTS {location} "
+            f"CREATE {table_type.upper()} TABLE IF NOT EXISTS {target_table_name} "
             f"({create_table_columns})"
             f" /* Python:snowflake.connector.pandas_tools.write_pandas() */ "
         )
         logger.debug(f"auto creating table with '{create_table_sql}'")
         cursor.execute(create_table_sql, _is_internal=True)
-        drop_file_format_sql = f"DROP FILE FORMAT IF EXISTS {file_format_name}"
-        logger.debug(f"dropping file format with '{drop_file_format_sql}'")
-        cursor.execute(drop_file_format_sql, _is_internal=True)
-
-    # in Snowflake, all parquet data is stored in a single column, $1, so we must select columns explicitly
-    # see (https://docs.snowflake.com/en/user-guide/script-data-load-transform-parquet.html)
-    if quote_identifiers:
-        parquet_columns = "$1:" + ",$1:".join(f'"{c}"' for c in df.columns)
     else:
-        parquet_columns = "$1:" + ",$1:".join(df.columns)
+        target_table_name = f"{location}{quote}{table_name}{quote}"
+
+    try:
+        copy_into_sql = (
+            f"COPY INTO {target_table_name} /* Python:snowflake.connector.pandas_tools.write_pandas() */ "
+            f"({columns}) "
+            f'FROM (SELECT {parquet_columns} FROM @"{stage_name}") '
+            f"FILE_FORMAT=(TYPE=PARQUET COMPRESSION={compression_map[compression]}) "
+            f"PURGE=TRUE ON_ERROR={on_error}"
+        )
+        logger.debug(f"copying into with '{copy_into_sql}'")
+        copy_results = cursor.execute(copy_into_sql, _is_internal=True).fetchall()
+
+        if overwrite:
+            original_table_name = f"{location}{quote}{table_name}{quote}"
+            drop_object(original_table_name, "table")
+            rename_table_sql = f"ALTER TABLE IF EXISTS {target_table_name} RENAME TO {original_table_name} /* Python:snowflake.connector.pandas_tools.write_pandas() */"
```

Comment:
This table needs to exists, let's remove the error catching (we should know if this ever happened)
```suggestion
            rename_table_sql = f"ALTER TABLE {target_table_name} RENAME TO {original_table_name} /* Python:snowflake.connector.pandas_tools.write_pandas() */"
```